File: algorithms/implementation/migratory-birds/migratory-birds-python/tests/migratoryBirds.py
# No problema "Migratory Birds" no HackerRank, 
# a tarefa é encontrar o tipo de pássaro mais 
# comum entre um grupo de pássaros migratórios. 
# Cada pássaro é representado por um número inteiro 
# e o objetivo é determinar qual número aparece com mais frequência no grupo.

# Descrição do Problema:

# Dado um array que representa os tipos de pássaros 
# avistados, onde cada elemento do array é um número 
# inteiro representando um tipo de pássaro, sua tarefa 
# é encontrar e imprimir o tipo de pássaro que é mais 
# comum. Se houver empate, escolha o menor número.

# # Função que encontra o tipo mais comum de pássaro
# def migratoryBirds(arr):
    
#     # Inicializa um dicionário para contar a frequência de cada tipo de pássaro
#     dictN = {}
    
#     # Loop para percorrer cada item no array
#     for item in arr:
        
#         # Verifica se o item já está no dicionário
#         if item in dictN:
            
#             # Se sim, incrementa a contagem
#             dictN[item] += 1
            
#         else:
            
#             # Se não, adiciona o item ao dicionário com contagem 1
#             dictN[item] = 1
    
#     # Inicializa variáveis para armazenar o tipo mais comum e sua contagem máxima
#     max_ = 0
#     numero = 0
    
#     # Loop para encontrar o tipo mais comum
#     for key, valor in dictN.items():
        
#         # Verifica se a contagem atual é maior que a contagem máxima
#         if(valor > max_):
            
#             # Se sim, atualiza o tipo mais comum e a contagem máxima
#             max_ = valor 
#             numero = key 
            
#         # Se houver empate, escolhe o menor número
#         elif (valor == max_):
            
#             if(key < numero):
                
#                 numero = key
    
#     # Retorna o tipo mais comum
#     return numero


# Função que encontra o tipo mais comum de pássaro

def migratoryBirds(arr):
    
    # Inicializa um dicionário para contar a frequência de cada tipo de pássaro
    dicionario = {}
    
    # Loop para percorrer cada tipo de pássaro no array
    for i in arr:
        
        # Verifica se o tipo de pássaro já está no dicionário
        if i not in dicionario.keys():
            
            # Se não, adiciona o tipo de pássaro ao dicionário com contagem 1
            dicionario[i] = 1
            
        else:
            
            # Se sim, incrementa a contagem
            dicionario[i] += 1
            
            
    # dicionario.values(): Obtém uma visão dos valores do dicionário dicionario. Neste caso, são as contagens de cada tipo de pássaro.

    # max(dicionario.values()): Encontra o valor máximo entre as contagens de tipos de pássaro no dicionário.

    # [k for k, v in dicionario.items() if v == max(dicionario.values())]: Cria uma lista de chaves (k) para as quais os valores (v) 
    # são iguais ao valor máximo encontrado. Ou seja, esta lista contém as chaves dos tipos de pássaros que têm a contagem máxima.

    # min(...): Encontra o valor mínimo dentro da lista criada no passo anterior. Isso significa que, em caso de empate, a menor chave (tipo de pássaro) será escolhida.
                
    # Retorna o tipo mais comum, escolhendo o menor número em caso de empate
    return min([k for k, v in dicionario.items() if v == max(dicionario.values())])

File: algorithms/implementation/migratory-birds/migratory-birds-javascript/js/migratoryBirds.js
// # No problema "Migratory Birds" no HackerRank, 
// # a tarefa é encontrar o tipo de pássaro mais 
// # comum entre um grupo de pássaros migratórios. 
// # Cada pássaro é representado por um número inteiro 
// # e o objetivo é determinar qual número aparece com mais frequência no grupo.

// # Descrição do Problema:

// # Dado um array que representa os tipos de pássaros 
// # avistados, onde cada elemento do array é um número 
// # inteiro representando um tipo de pássaro, sua tarefa 
// # é encontrar e imprimir o tipo de pássaro que é mais 
// # comum. Se houver empate, escolha o menor número.


// Função que encontra o tipo mais comum de pássaro

// export function migratoryBirds(arr) {

//     // Cria um mapa para contar a frequência de cada tipo de pássaro
//     let contar_passaro = new Map();

//     // Loop para percorrer cada tipo de pássaro no array
//     for (const tipoPassaro of arr) {

//         // Verifica se o tipo de pássaro já está no mapa
//         if (contar_passaro.has(tipoPassaro)) {
            
//             // Se sim, incrementa a contagem
//             contar_passaro.set(tipoPassaro, contar_passaro.get(tipoPassaro) + 1);

//         } else {

//             // Se não, adiciona o tipo de pássaro ao mapa com contagem 1
//             contar_passaro.set(tipoPassaro, 1);
//         }
//     }
    
//     // Inicializa variáveis para armazenar os tipos mais frequentes e a contagem máxima
//     let avesMaisFreqüentes = [];
//     let maxContagem = 0;

//     // Loop sobre o mapa para encontrar os tipos mais frequentes
//     contar_passaro.forEach((contar, tipoPassaro) => {

//         // Verifica se a contagem atual é maior que a contagem máxima
//         if (contar > maxContagem) {
            
//             // Se sim, atualiza os tipos mais frequentes e a contagem máxima
//             maxContagem = contar;
//             avesMaisFreqüentes = [tipoPassaro];

//         } else if (contar === maxContagem) {

//             // Se houver empate, adiciona o tipo ao array de tipos mais frequentes
//             avesMaisFreqüentes.push(tipoPassaro);
            
//         }
//     });

//     // Retorna o tipo mais comum, escolhendo o menor número em caso de empate
//     return Math.min(...avesMaisFreqüentes);
// }

export function migratoryBirds(arr) {

    // Inicializa um array para armazenar a contagem de cada tipo de pássaro (inicialmente todos são 0)
    let tipoDePassaroMaximo = [0, 0, 0, 0, 0];

    // Itera sobre o array de observações de pássaros
    for (const passaro of arr) {
        
        // Usa um switch para incrementar a contagem do tipo de pássaro correspondente
        switch (passaro) {

            case 1:
                
            tipoDePassaroMaximo[0]++
                break;

            case 2:
            
            tipoDePassaroMaximo[1]++
                break;

            case 3:
        
            tipoDePassaroMaximo[2]++
                break;

            case 4:
        
            tipoDePassaroMaximo[3]++
                break;

            case 5:
        
            tipoDePassaroMaximo[4]++
                break;
        }
    }

    // Inicializa a variável resultado com o tipo de pássaro mais comum (inicialmente 1)
    let resultado = 1;

    // Itera sobre o array de contagem para encontrar o tipo de pássaro mais comum
    for (let index = 1; index < tipoDePassaroMaximo.length; index++) {
        
        // Compara a contagem atual com a contagem do tipo de pássaro mais comum até agora
        if (tipoDePassaroMaximo[index] > tipoDePassaroMaximo[resultado - 1]) {
            
            // Atualiza a variável resultado com o novo tipo de pássaro mais comum
            resultado = index + 1;
        }
        
    }

    // Retorna o tipo de pássaro mais comum
    return resultado;
}